doc/cfkey.8
.TH cfengine 8 "Maintenance Commands" HiOslo
.SH NAME
cfengine \- network configuration engine
.SH SYNOPSIS

/usr/local/sbin/cfkey


.SH DESCRIPTION
.B cfengine
is a language based system specifically designed
for testing and configuring unix\-like systems attached to
a TCP/IP network. The command generates a pair of public-private
keys for remote authentication during cfrun or cfagent and 
the server cfservd.

.SH AUTHOR
Mark Burgess, Oslo University College
.SH INFORMATION
http://www.cfengine.org



ChangeLog
2.2.10

	Protocol fix, potential integer error in server decoding.

2.2.9
	
2.2.8
	
        Patch for cfservd serving wrong data when link->copy

	Checksum changes now trigger persistent class "checksum_alerts"
	and log to WORKDIR/state/file_hash_event_history

	Threading error fix for cfexecd

	Windows compilation fixes. Note that windows users might have to move keys and files
	from /home/user/.cfagent to /var/cfengine after the standardization of administrator
	paths for windows where getuid != 0.
	
2.2.7
	Memory leak in database access, affecting cfenvd
	
2.2.6

	Xen detection
	
	Bug in popen - incorrect placement of conditional in setting groups added in 2.2.4.

	Sensor code added for linux cfenvd. Ready for patches for other systems.
	Cpu utilization and temperature measurement added for linux where available

	Bug reported in installable-negated classes. Patch by jack/SiliconSlick

	Bugfix for class defined by directories
	
2.2.5
	No content changes since 2.2.4, just fixes a packaging error.
	(Mark's Easter turkey)
2.2.4p
	Blunder in makefile change. Compilation error fixed.
	
2.2.4	
	Build changes. libpub is now an installed library libcfpub.so/a
	so that cfengine 3 can find it.

	Documentation is no longer distributed in the tarball, since most
	people will use the online version. Also, enhanced documentation is
	now available for any user who registers (free) at cfengine.com.

	compress=true option in tidy was missing from documentation, and
	left uninitialized during purging leading to attempts to compress
	files instead of deleting. 

	Bugfix for empty strings in editfiles.

	Links did not get removed with nolink=kill. Code changes apparently broken earlier.
	Fixed.

	rxdirs= added to files. Default is false. If set to true, cfengine will not
	assume that x-bits should be set when r is set on directory permissions.

	Group initialization to popen, so shellcommands groups are consistent with uid login id

	Rpm fix for package removal.
	
2.2.3
	Feedback release after 2.2.2 and LISA 07, minor tweaks and bug fixes

	"Harmless features"
	
	IgnoreInterfaceRegex( "vm.*" ) added to control
	function added for removing data from special interfaces from class data

	Option checkroot=true/false added to copy directive. This switches
	on/off the change in 2.2.2 about setting permissions on the
	destination root when recursively copying files. Default is now
	false.

	Can escape colons in the strings now to avoid confusion with
	expansion separator. 

	cfenvgraph --erasehistory users,www_in etc will now set the
	average and variance of learned cfenvd data to zero. The current
	week's trace is not affected.

	audit=true/false flag added to most actions
	
2.2.2
	
	The build procedure for cfengine changes in this version to
	incorporate an intermediate library. The default behaviour is now
	to build a static library which will be used by cfengine 3 and
	other application interfaces. To build a shared library, use

	--enable-shared=yes

	NOTE to developers: if running from the compiled code you now need
	to delete the files cfagent cfservd etc and recompile to see changes
	appear in the code. This is due to the libtool build process.
	
	libtoolizing compilation process for migration to cfengine shared library
	(this is to support mulitple interfaces in future, and cf 3)

	--

	XML and HTML output options added to equip the internal instrumentation 
	with readable output in cfshow e.g.

	   cfshow --performance --html > output.html
	   firefox output.html

	provides an easier summary and searchability in browser function

	Instrumentation changed to use realtime functions where possible. Standard
	unix timer is too insensitive to measure performance.

	The beginning of an integrated auditing system has been added. See references
	to AuditLog() in the code. This logs to a compressed database and can be extracted
	in ascii, html, xml using cfshow, e.g. cfshow  --audit --html > output.html
	Auditing is switched on by

	control:
	 cfinputs_version = ( 1.2.1 ) # record this version in log, e.g. cvs revision
	 Auditing = true
	
	--

	cfshow --classes 

	also shows the relative frequencies of classes
	that have occurred on the system and the last time these were
	observed.

	cfshow -f cfagent.conf --regex "lin.*"

	displays all rules in the named file whose classes match the named
	regular expression. Note that the class "any" is not automatically matched
	and the search is based on the class text from the file. The output is not
	related to which classes are currently defined, and it will not recognise
	the history of classes implicit through class-restricted imported files.
	
	--
	
	SCLI interface modifications.

	--
	
	Intermittency entropy code added to the copy instrumentation for
	automatic host reliability measurements.

	Special variable

	   IgnoreFriendRegex = ( regex )

	prevents last-seen warnings about hosts matching the named regex
	
	--
	
	UserExists("name") and GroupExists("name") functions added to builtin functions.

	ClassMatch function added in classes/groups to match regular expressions of other classes

	-- 
	
	Additional octet variables added on default IPv4 address, for convenience:
	e.g.
	   590  : ipv4_2[eth1]=192.168
	   1895 : ipv4[eth1]=192.168.1.101
	   2100 : ipv4_1[eth1]=192
	   4049 : ipv4_3[eth1]=192.168.1
	

	Methods can now expand some list variables, one list only per argument set.
	This is a limited compromise pending cfengine 3. A patch for expansion of
	variables in the control part of a program in methods has also been added.

	Bug fix and code simplication of variable expansion.

	-----

	tied variable expansion added. Allows method workaround for file dissemination
	by name, e.g. copy from

	host1.conf -> /dir/host1/file.conf
        host2.conf -> /dir/host2/file.conf  e.g. for virtual or diskless machines
	
	-----

	Major changes to package manager code by Jo Rhett. Variable expansion
	fixed by Mark.

	-----


src/cfkey.c
/* 

        Copyright (C) 1999
        Free Software Foundation, Inc.

   This file is part of GNU cfengine - written and maintained 
   by Mark Burgess, Dept of Computing and Engineering, Oslo College,
   Dept. of Theoretical physics, University of Oslo
 
   This program is free software; you can redistribute it and/or modify it
   under the terms of the GNU General Public License as published by the
   Free Software Foundation; either version 2, or (at your option) any
   later version.
 
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
 
  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA

*/

/*****************************************************************************/
/*                                                                           */
/* File: cfkey.c                                                             */
/*                                                                           */
/*****************************************************************************/

#include "cf.defs.h"
#include "cf.extern.h"

char CFLOCK[CF_BUFSIZE];
char USERKEYFILE[CF_BUFSIZE];

void Initialize(void);
int RecursiveTidySpecialArea(char *name, struct Tidy *tp, int maxrecurse, struct stat *sb);
void Syntax(void);
void CheckOpts(int argc, char **argv);

/*******************************************************************/
/* Command line options                                            */
/*******************************************************************/

  /* GNU STUFF FOR LATER #include "getopt.h" */
 
 struct option OPTIONS[5] =
      {
      { "help",no_argument,0,'h' },
      { "file",required_argument,0,'f' },
      { "version",no_argument,0,'V' },
      { "quiet",no_argument,0,'w' },
      { NULL,0,0,0 }
      };


int main(int argc, char** argv)

{ unsigned long err;
  RSA *pair;
  FILE *fp;
  struct stat statbuf;
  int fd;
  static char *passphrase = "Cfengine passphrase";
  EVP_CIPHER *cipher = EVP_des_ede3_cbc();

CheckOpts(argc, argv);

Initialize();
           
if (stat(CFPRIVKEYFILE,&statbuf) != -1)
   {
   printf("A key file already exists at %s.\n",CFPRIVKEYFILE);
   return 1;
   }

if (stat(CFPUBKEYFILE,&statbuf) != -1)
   {
   printf("A key file already exists at %s.\n",CFPUBKEYFILE);
   return 1;
   }

 if(VERBOSE) 
   {
   printf("Making a key pair for cfengine, please wait, this could take a minute...\n"); 
   }

pair = RSA_generate_key(2048,35,NULL,NULL);

if (pair == NULL)
   {
   err = ERR_get_error();
   printf("Error = %s\n",ERR_reason_error_string(err));
   return 1;
   }

if (VERBOSE)
   {
   RSA_print_fp(stdout,pair,0);
   }
 
fd = open(CFPRIVKEYFILE,O_WRONLY | O_CREAT | O_TRUNC, 0600);

if (fd < 0)
   {
   printf("Ppen %s failed: %s.",CFPRIVKEYFILE,strerror(errno));
   return 1;
   }
 
if ((fp = fdopen(fd, "w")) == NULL )
   {
   printf("fdopen %s failed: %s.",CFPRIVKEYFILE, strerror(errno));
   close(fd);
   return 1;
   }

if (VERBOSE)
   {
printf("Writing private key to %s\n",CFPRIVKEYFILE);
   }
 
if (!PEM_write_RSAPrivateKey(fp,pair,cipher,passphrase,strlen(passphrase),NULL,NULL))
    {
    err = ERR_get_error();
    printf("Error = %s\n",ERR_reason_error_string(err));
    return 1;
    }
 
fclose(fp);
 
fd = open(CFPUBKEYFILE,O_WRONLY | O_CREAT | O_TRUNC, 0600);

if (fd < 0)
   {
   printf("open %s failed: %s.",CFPUBKEYFILE,strerror(errno));
   return 1;
   }
 
if ((fp = fdopen(fd, "w")) == NULL )
   {
   printf("fdopen %s failed: %s.",CFPUBKEYFILE, strerror(errno));
   close(fd);
   return 1;
   }

if (VERBOSE)
   {
   printf("Writing public key to %s\n",CFPUBKEYFILE);
   }
 
if(!PEM_write_RSAPublicKey(fp,pair))
   {
   err = ERR_get_error();
   printf("Error = %s\n",ERR_reason_error_string(err));
   return 1;
   }

fclose(fp);

 
snprintf(VBUFF,CF_BUFSIZE,"%s/randseed",VLOGDIR);
RAND_write_file(VBUFF);
chmod(VBUFF,0644); 
return 0;
}

/*******************************************************************/
/* Level 1                                                         */
/*******************************************************************/

void Initialize()

{
umask(077);
 /* XXX Initialize workdir for non privileged users */

strcpy(CFWORKDIR,WORKDIR);

#ifndef NT
if (geteuid() != 0)
   {
   char *homedir;
   if ((homedir = getenv("HOME")) != NULL)
      {
      strcpy(CFWORKDIR,homedir);
      strcat(CFWORKDIR,"/.cfagent");
      }
   }
#endif
 
strcpy(VLOCKDIR,CFWORKDIR);
strcpy(VLOGDIR,CFWORKDIR); 

OpenSSL_add_all_algorithms();
ERR_load_crypto_strings();

CheckWorkDirectories();
if (USERKEYFILE[0] != '\0')
  {
    strncpy(CFPRIVKEYFILE, USERKEYFILE, CF_BUFSIZE-6);
    strcat(CFPRIVKEYFILE, ".priv");
    strncpy(CFPUBKEYFILE, USERKEYFILE, CF_BUFSIZE-5);
    strcat(CFPUBKEYFILE, ".pub");
  }
RandomSeed();
}

/*******************************************************************/

void Syntax()

{ int i;

printf("GNU cfengine: A system configuration engine (cfkey)\n%s\n%s\n",VERSION,COPYRIGHT);
printf("\n");
printf("Options:\n\n");

for (i=0; OPTIONS[i].name != NULL; i++)
   {
   printf("--%-20s    (-%c)\n",OPTIONS[i].name,(char)OPTIONS[i].val);
   }

printf("\nBug reports to bug-cfengine@cfengine.org\n");
printf("General help to help-cfengine@cfengine.org\n");
printf("Info & fixes at http://www.cfengine.org\n");
}

/*****************************************************************************/

void CheckOpts(int argc,char **argv)

{ extern char *optarg;
  struct Item *actionList;
  char *tmp, *suffix;
  int optindex = 0;
  int c;

  while ((c=getopt_long(argc,argv,"f:hvVw",OPTIONS,&optindex)) != EOF)
  {
  switch ((char) c)
      {
      case 'f':
          strncpy(USERKEYFILE, optarg, CF_BUFSIZE-1);
          /* search for .priv or .pub suffix and wipe it out */
          suffix = strrchr(USERKEYFILE, '.');
          if (NULL != suffix 
              && ((0 == strcmp(suffix, ".priv")) 
                  || (0 == strcmp(suffix, ".pub"))))
            *suffix = '\0';
          break;
          
      case 'V': printf("GNU cfengine %s\n%s\n",VERSION,COPYRIGHT);
          printf("This program is covered by the GNU Public License and may be\n");
          printf("copied free of charge.  No warranty is implied.\n\n");
          exit(0);
          
      case 'h': Syntax();
          exit(0);

      case 'w': VERBOSE = false;
           break;
          
      default:  Syntax();
          exit(1);
         
      }
  }
}

